Add support for np.ndarray in ragged tensor construction.

PhilJd opened the pull request:
First of all, thanks a lot for implementing the awesome ragged tensors! :)

This pull request adds support for `np.ndarray ` as nested type. With the pull request, the following is possible:
```
tf.ragged.constant([np.array([1,2]), np.array([4,5,6])])
```
With the current head, this snippet throws:
`TypeError: Eager execution of tf.constant with unsupported shape (value has 6 elements, shape is (3,) with 3 elements)`

I guess storing data as lists of (differently sized)  numpy arrays is not too uncommon, e.g., in object detection there's usually a different number of bounding boxes per image.

Note: I couldn't find a test file for `ragged_factory_op`, where should unit tests for this go?

PhilJd committed:
Add support for np.ndarray in ragged tensor construction.

alextp reviewed (changes requested):
Can you add tests?

Otherwise this looks great!

PhilJd commented:
Yep, I intended to, but I couldn't find a test file for `ragged_factory_op`. Should I create a new file `ragged_factory_op_test.py` or does it fit into one of the other test files?

alextp commented:
Up to you.

On Tue, Feb 19, 2019 at 1:26 PM Phil <notifications@github.com> wrote:

> Yep, I intended to, but I couldn't find a test file for ragged_factory_op.
> Should I create a new file ragged_factory_op_test.py or does it fit into
> one of the other test files?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/25849#issuecomment-465315469>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAATxV5Hl2cab7pYEgptn_LzvUjL3tFkks5vPGv4gaJpZM4bBkSs>
> .
>


-- 
 - Alex

rthadur commented:
@PhilJd gentle ping to update tests

PhilJd committed:
Switch scalar check in ragged tensors from using isinstance to np.ndim to allow for scalar np.ndarrays. Add test cases for construction of ragged tensors from nested numpy arrays.

PhilJd commented:
Hi,
I'm extremely sorry for the long delay, thanks for pinging!

I've added tests and fixed the code to handle scalar np.arrays correctly.
Specifically, I've replaced `isinstance(item, (list, tuple, np.ndarray))` which verifies that `item` is not a scalar with `np.ndim(item) != 0`.

On a side note: My machine just broke and my temporary machine is not fully setup. I only ran the ragged test cases but not the complete tensorflow test suite as I would normally do.

Again, sorry for the delay!
Cheers,
Phil

edloper commented on tensorflow/python/ops/ragged/ragged_factory_ops.py:
```diff
@@ -155,7 +156,7 @@ def _constant_value(ragged_factory, inner_factory, pylist, dtype, ragged_rank,
     raise TypeError("pylist may not be a RaggedTensor or RaggedTensorValue.")
 
-  if not isinstance(pylist, (list, tuple)):
+  if np.ndim(pylist) == 0:
```
It looks to me like np.ndim scans the entire input, which could be unnecessarily expensive.  How about changing this to:

    if np.isscalar(pylist) or not isinstance(pylist, (list, tuple, np.ndarray):

PhilJd replied:
The problem with this is that `np.isscalar(np.ndarray(1))` evaluates to `False`. How about short-circuiting `list` and `tuple`?
```
if not isinstance(pylist, (list, tuple)) and np.ndim(pylist) == 0:
```

edloper replied:
Sounds good to me. Thanks.

PhilJd committed:
Short-circuit np.dim evaluation for list and tuple.
